Implement Wasm tail calls in IPInt

danlliu opened the pull request:
#### 34f187fe40afbd66733d33ebb514da0ae2a40177
<pre>
Implement Wasm tail calls in IPInt
<a href="https://bugs.webkit.org/show_bug.cgi?id=281785">https://bugs.webkit.org/show_bug.cgi?id=281785</a>
<a href="https://rdar.apple.com/138213801">rdar://138213801</a>

Reviewed by Keith Miller.

This patch implements `return_call` and `return_call_indirect` in IPInt, bringing
us closer to feature parity with LLInt.

* JSTests/wasm/ipint-tests/ipint-test-call-indirect-arg.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.type.i2i.func.param.i32.result.i32.table.table.export.string_appeared_here.1.funcref.elem.i32.const.0.inc.func.inc.export.string_appeared_here.param.i32.result.i32.local.0.i32.const.1.i32.add.return.func.export.string_appeared_here.param.i32.result.i32.i32.const.1094795585.local.0.i32.const.0.call_indirect.table.type.i2i.i32.const.0.call_indirect.table.type.i2i.return.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-indirect-simple.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.type.sig.func.param.i64.param.i64.result.i64.table.table.export.string_appeared_here.1.funcref.elem.i32.const.0.fac_aux.func.fac.export.string_appeared_here.param.x.i64.result.i64.return_call.fac_aux.local.x.i64.const.1.func.fac_aux.param.x.i64.param.r.i64.result.i64.result.i64.i64.eqz.local.x.then.return.local.r.else.i64.sub.local.x.i64.const.1.i64.mul.local.x.local.r.i32.const.0.return_call_indirect.table.type.sig.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-simple.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.fac.export.string_appeared_here.param.x.i64.result.i64.return_call.fac_aux.local.x.i64.const.1.func.fac_aux.param.x.i64.param.r.i64.result.i64.result.i64.i64.eqz.local.x.then.return.local.r.else.i64.sub.local.x.i64.const.1.i64.mul.local.x.local.r.return_call.fac_aux.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-stack-bigchange.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.test.export.string_appeared_here.param.i32.result.i32.i32.const.1.i32.const.2.i32.const.3.i32.const.4.i32.const.5.i32.const.6.i32.const.7.i32.const.8.i32.const.9.i32.const.10.i32.const.11.i32.const.12.call.caller.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.func.caller.result.i32.return_call.callee.i32.const.21.i32.const.22.i32.const.23.i32.const.24.i32.const.25.i32.const.26.i32.const.27.i32.const.28.i32.const.29.i32.const.30.i32.const.31.i32.const.32.i32.const.33.i32.const.34.i32.const.35.i32.const.36.func.callee.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.result.i32.i32.sub.local.15.local.0.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-stack-params-less.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.test.export.string_appeared_here.param.i32.result.i32.i32.const.1.i32.const.2.i32.const.3.i32.const.4.i32.const.5.i32.const.6.i32.const.7.i32.const.8.i32.const.9.i32.const.10.i32.const.11.i32.const.12.call.caller.i32.const.20.i32.const.21.i32.const.22.i32.const.23.i32.const.24.i32.const.25.i32.const.26.i32.const.27.i32.const.28.i32.const.29.i32.const.30.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.func.caller.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.result.i32.return_call.callee.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.func.callee.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.result.i32.i32.const.40.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-stack-params-more.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.test.export.string_appeared_here.param.i32.result.i32.i32.const.1.i32.const.2.i32.const.3.i32.const.4.i32.const.5.i32.const.6.i32.const.7.i32.const.8.i32.const.9.i32.const.10.i32.const.11.i32.const.12.call.caller.i32.const.20.i32.const.21.i32.const.22.i32.const.23.i32.const.24.i32.const.25.i32.const.26.i32.const.27.i32.const.28.i32.const.29.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.func.caller.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.result.i32.return_call.callee.i32.const.30.i32.const.31.i32.const.32.i32.const.33.i32.const.34.i32.const.35.i32.const.36.i32.const.37.i32.const.38.i32.const.39.i32.const.40.i32.const.41.func.callee.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.result.i32.i32.const.50.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-stack-params-same.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.test.export.string_appeared_here.param.i32.result.i32.i32.const.1.i32.const.2.i32.const.3.i32.const.4.i32.const.5.i32.const.6.i32.const.7.i32.const.8.i32.const.9.i32.const.10.i32.const.11.i32.const.12.call.caller.i32.const.20.i32.const.21.i32.const.22.i32.const.23.i32.const.24.i32.const.25.i32.const.26.i32.const.27.i32.const.28.i32.const.29.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.func.caller.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.result.i32.return_call.callee.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.local.9.func.callee.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.result.i32.i32.const.30.async test):
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/llint/InPlaceInterpreter.h:
* Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
* Source/JavaScriptCore/llint/LLIntData.cpp:
(JSC::LLInt::initialize):
* Source/JavaScriptCore/runtime/Gate.h:
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::IPIntGenerator):
(JSC::Wasm::IPIntGenerator::addCallCommonData):
(JSC::Wasm::IPIntGenerator::addTailCallCommonData):
(JSC::Wasm::IPIntGenerator::addCall):
(JSC::Wasm::IPIntGenerator::addCallIndirect):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.h:
* Tools/lldb/debug_ipint.py:

Canonical link: <a href="https://commits.webkit.org/286874@main">https://commits.webkit.org/286874@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8bbafb5dd4194d86ec1e5c3cc48d55b9ea05e505

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77173 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56208 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30088 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81734 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28455 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79290 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65356 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4504 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60473 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18521 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80240 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50427 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66239 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40775 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47829 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23737 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26778 "Built successfully") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/70358 "Found 3 new JSC stress test failures: stress/get-by-val-hoist-above-structure.js.ftl-no-cjit-no-inline-validate, stress/get-by-val-hoist-above-structure.js.mini-mode, wasm.yaml/wasm/modules/wasm-imports-js-exports.js.wasm-no-cjit (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68944 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24072 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83160 "Built successfully") | 
| [❌ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/76451 "Found 2 new JSC stress test failures: stress/attribute-custom-accessor.js.dfg-eager, wasm.yaml/wasm/stress/referenced-function.js.wasm-collect-continuously (failure)") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4553 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3065 "Found 1 new test failure: webrtc/vp8-then-h264.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68746 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4709 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66212 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68001 "Found 2 new API test failures: /TestWebKit:WebKit.LoadAlternateHTMLStringWithNonDirectoryURL, /TestWebKit:WebKit.LoadAlternateHTMLStringWithEmptyBaseURL (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11983 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10075 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/98704 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11974 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4500 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21593 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4519 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7954 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6278 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -521,8 +521,52 @@ instructionLabel(_call_indirect)
+    subq 16, sp
+    move sp, a2
+
+    # operation returns the entrypoint in r0 and the target instance in r1
```
/# operation returns the entrypoint in r0 and the target instance in r1/# operation returns the entrypoint in r0, the target instance in r1, and writes to sp the boxedCallee/s

As an aside, maybe it would be clearer if we returned the target callee in r0 and loaded/boxed the callee in asm instead of writing to the stack?

kmiller68 commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -521,8 +521,52 @@ instructionLabel(_call_indirect)
+    move sp, a2
+
+    # operation returns the entrypoint in r0 and the target instance in r1
+    operationCall(macro() cCall3(_ipint_extern_call) end)
```
Nit: Maybe we should rename this to `ipint_extern_prepare_call`? Since it doesn't actually do the call.

kmiller68 commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -521,8 +521,52 @@ instructionLabel(_call_indirect)
+    # Get callIndirectMetadata
+    move cfr, a1
+    move MC, a3
+    operationCall(macro() cCall4(_ipint_extern_call_indirect) end)
```
Nit: Ditto to `ipint_extern_prepare_call_indirect`?

kmiller68 commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -5072,7 +5116,7 @@ end
     loadb [MC], sc0
     addq 1, MC
-    bilt sc0, 20, .safe
+    bilt sc0, 24, .safe
```
Can we make this a `(constexpr NumMIntArgOpcodes)`?

kmiller68 reviewed (approved):
r=me with some comments.

justinmichaud commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -5055,7 +5100,7 @@ slowPathLabel(_local_tee)
 
 # time to use the safe for call registers!
-# sc0 = mINT shadow stack pointer (tracks the Wasm stack)
+# sc1 = mINT shadow stack pointer (tracks the Wasm stack)
```
Can we give this a name?

danlliu replied:
It's named `mintSS` underneath; I just had to update the comment since I was reallocating it to `sc1`.

danlliu committed:
Implement Wasm tail calls in IPInt
https://bugs.webkit.org/show_bug.cgi?id=281785
rdar://138213801

Reviewed by Keith Miller.

This patch implements `return_call` and `return_call_indirect` in IPInt, bringing
us closer to feature parity with LLInt.

* JSTests/wasm/ipint-tests/ipint-test-call-indirect-arg.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.type.i2i.func.param.i32.result.i32.table.table.export.string_appeared_here.1.funcref.elem.i32.const.0.inc.func.inc.export.string_appeared_here.param.i32.result.i32.local.0.i32.const.1.i32.add.return.func.export.string_appeared_here.param.i32.result.i32.i32.const.1094795585.local.0.i32.const.0.call_indirect.table.type.i2i.i32.const.0.call_indirect.table.type.i2i.return.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-indirect-simple.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.type.sig.func.param.i64.param.i64.result.i64.table.table.export.string_appeared_here.1.funcref.elem.i32.const.0.fac_aux.func.fac.export.string_appeared_here.param.x.i64.result.i64.return_call.fac_aux.local.x.i64.const.1.func.fac_aux.param.x.i64.param.r.i64.result.i64.result.i64.i64.eqz.local.x.then.return.local.r.else.i64.sub.local.x.i64.const.1.i64.mul.local.x.local.r.i32.const.0.return_call_indirect.table.type.sig.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-simple.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.fac.export.string_appeared_here.param.x.i64.result.i64.return_call.fac_aux.local.x.i64.const.1.func.fac_aux.param.x.i64.param.r.i64.result.i64.result.i64.i64.eqz.local.x.then.return.local.r.else.i64.sub.local.x.i64.const.1.i64.mul.local.x.local.r.return_call.fac_aux.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-stack-bigchange.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.test.export.string_appeared_here.param.i32.result.i32.i32.const.1.i32.const.2.i32.const.3.i32.const.4.i32.const.5.i32.const.6.i32.const.7.i32.const.8.i32.const.9.i32.const.10.i32.const.11.i32.const.12.call.caller.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.func.caller.result.i32.return_call.callee.i32.const.21.i32.const.22.i32.const.23.i32.const.24.i32.const.25.i32.const.26.i32.const.27.i32.const.28.i32.const.29.i32.const.30.i32.const.31.i32.const.32.i32.const.33.i32.const.34.i32.const.35.i32.const.36.func.callee.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.result.i32.i32.sub.local.15.local.0.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-stack-params-less.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.test.export.string_appeared_here.param.i32.result.i32.i32.const.1.i32.const.2.i32.const.3.i32.const.4.i32.const.5.i32.const.6.i32.const.7.i32.const.8.i32.const.9.i32.const.10.i32.const.11.i32.const.12.call.caller.i32.const.20.i32.const.21.i32.const.22.i32.const.23.i32.const.24.i32.const.25.i32.const.26.i32.const.27.i32.const.28.i32.const.29.i32.const.30.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.func.caller.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.result.i32.return_call.callee.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.func.callee.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.result.i32.i32.const.40.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-stack-params-more.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.test.export.string_appeared_here.param.i32.result.i32.i32.const.1.i32.const.2.i32.const.3.i32.const.4.i32.const.5.i32.const.6.i32.const.7.i32.const.8.i32.const.9.i32.const.10.i32.const.11.i32.const.12.call.caller.i32.const.20.i32.const.21.i32.const.22.i32.const.23.i32.const.24.i32.const.25.i32.const.26.i32.const.27.i32.const.28.i32.const.29.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.func.caller.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.result.i32.return_call.callee.i32.const.30.i32.const.31.i32.const.32.i32.const.33.i32.const.34.i32.const.35.i32.const.36.i32.const.37.i32.const.38.i32.const.39.i32.const.40.i32.const.41.func.callee.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.result.i32.i32.const.50.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-stack-params-same.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.test.export.string_appeared_here.param.i32.result.i32.i32.const.1.i32.const.2.i32.const.3.i32.const.4.i32.const.5.i32.const.6.i32.const.7.i32.const.8.i32.const.9.i32.const.10.i32.const.11.i32.const.12.call.caller.i32.const.20.i32.const.21.i32.const.22.i32.const.23.i32.const.24.i32.const.25.i32.const.26.i32.const.27.i32.const.28.i32.const.29.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.func.caller.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.result.i32.return_call.callee.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.local.9.func.callee.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.param.i32.result.i32.i32.const.30.async test):
* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/llint/InPlaceInterpreter.h:
* Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
* Source/JavaScriptCore/llint/LLIntData.cpp:
(JSC::LLInt::initialize):
* Source/JavaScriptCore/runtime/Gate.h:
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::IPIntGenerator):
(JSC::Wasm::IPIntGenerator::addCallCommonData):
(JSC::Wasm::IPIntGenerator::addTailCallCommonData):
(JSC::Wasm::IPIntGenerator::addCall):
(JSC::Wasm::IPIntGenerator::addCallIndirect):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.h:
* Tools/lldb/debug_ipint.py:

Canonical link: https://commits.webkit.org/286874@main
